### PR TITLE
Convert the complexity analyzer into a validation rule

### PR DESCRIPTION
**Pull request by Shane32**
This PR focuses on removing the complexity analyzer from ExecutionOptions and instead triggering it from a validation rule.

This PR does not rewrite the code within the complexity analyzer.  Performance should be approximately the same as without this PR.

None of the builder methods have changed.

### CODE REVIEW

**Comment by Shane32** [comment]
Todo: restore builder method tests; add migration notes

File: docs2/site/docs/migrations/migration6.md [hunk 1]
```diff
@@ -24,3 +24,84 @@ Removal was necessary as the value is released after the result is set.
 
 Note that `MatchingNodeVisitor` has not changed, so many validation rules will not require
 any source code changes.
+
+### 3. `ExecutionOptions.ComplexityConfiguration` has been removed
+
+Complexity analysis is now a validation rule and has been removed from execution options.
+There is no change when using the `IGraphQLBuilder.AddComplexityAnalyzer` methods as shown below:
+
+```cs
```
Comment (sungam3r) [inline docs2/site/docs/migrations/migration6.md:33]:
```suggestion
```csharp
```

everywhere

File: docs2/site/docs/migrations/migration6.md [hunk 2]
```diff
@@ -24,3 +24,84 @@ Removal was necessary as the value is released after the result is set.
...
+
+// GraphQL 6.x
+options.ValidationRules = GraphQL.Validation.DocumentValidator.CoreRules.Append(
+    new ComplexityValidationRule(
+        complexityConfig,
+        options.RequestServices.GetRequiredService<IComplexityAnalyzer>()
+    ));
+```
+
+Using the `IGraphQLBuilder` interface to configure GraphQL is the recommended approach.
```
Comment (sungam3r) [inline docs2/site/docs/migrations/migration6.md:107]:
```suggestion
Using the `IGraphQLBuilder` interface to configure GraphQL execution engine is the recommended approach.
```

File: src/GraphQL/Validation/Complexity/ComplexityError.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+namespace GraphQL.Validation.Complexity;
+
+/// <summary>
+/// Represents a complexity error.
+/// </summary>
+public class ComplexityError : ValidationError
```
Comment (sungam3r) [inline src/GraphQL/Validation/Complexity/ComplexityError.cs:6]: 🤔 Or place it in Errors.Custom and new rule in Rules.Custom.

File: src/GraphQL/Validation/Complexity/ComplexityValidationRule.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        _complexityConfiguration = complexityConfiguration;
+        _complexityAnalyzer = complexityAnalyzer;
+    }
+
+    /// <inheritdoc/>
+    public ValueTask<INodeVisitor?> ValidateAsync(ValidationContext context)
+    {
+        try
+        {
+            _complexityAnalyzer.Validate(context.Document, _complexityConfiguration, context.Schema);
```
Comment (sungam3r) [inline src/GraphQL/Validation/Complexity/ComplexityValidationRule.cs:33]: I suggest to change Validate signature to return `ComplexityError(s)`.

File: src/GraphQL/Execution/DocumentExecuter.cs [hunk 5]
```diff
@@ -204,12 +196,6 @@ private async Task<ExecutionResult> CoreExecuteAsync(ExecutionOptions options)
                         }).ConfigureAwait(false);
                 }
 
-                if (options.ComplexityConfiguration != null && validationResult.IsValid && analyzeComplexity)
-                {
-                    using (metrics.Subject("document", "Analyzing complexity"))
```
Comment (sungam3r) [inline src/GraphQL/Execution/DocumentExecuter.cs:209]: @Shane32 What do you think about exposing metrics to all validation rules? Not a whole `Metrics` instance but some sort of method to delegate control flow to `public Marker Subject(string category, string? subject, Dictionary<string, object?>? metadata = null)`. Then we may preserve complexity time metrics from the removed code.

Comment (sungam3r) [reply src/GraphQL/Execution/DocumentExecuter.cs:209]: Or add `IMetrics` interface with single method `Marker Subject(string category, string? subject, Dictionary<string, object?>? metadata = null)` and add `IMetrics Metrics { get; set; }` property into `ValidationContext`.

**Review by sungam3r — APPROVED** [review]
@Shane32 Suggestions about further refactoring may be implemented in other PRs if you like but don't forget to convert them into issues. Also ping @SlavaUtesinov .

File: src/GraphQL/Validation/Complexity/ComplexityValidationRule.cs (hunk 4 above)
Comment (Shane32) [reply src/GraphQL/Validation/Complexity/ComplexityValidationRule.cs:33]: #3175

File: src/GraphQL/Execution/DocumentExecuter.cs (hunk 5 above)
Comment (Shane32) [reply src/GraphQL/Execution/DocumentExecuter.cs:209]: #3176

**Pull request merged**